src/hotspot/share/logging/logMessageBuffer.cpp

```diff
@@ -110,8 +110,14 @@ void LogMessageBuffer::vwrite(LogLevelType level, const char* fmt, va_list args)
 
     va_list copy;
     va_copy(copy, args);
-    written += (size_t)os::vsnprintf(current_buffer_position, remaining_buffer_length, fmt, copy) + 1;
+    int ret = os::vsnprintf(current_buffer_position, remaining_buffer_length, fmt, copy);
     va_end(copy);
+    assert(ret >= 0, "Log message buffer issue.");
```

Comment:
According to https://www.cplusplus.com/reference/cstdio/vsnprintf/ , the only reason for vsnprintf returning a negative value is when there's an encoding error.

Here's an example
https://stackoverflow.com/questions/28096802/what-is-the-meaning-of-encoding-error-in-the-vsnprintf-documentation

@jdksjolen have you seen an actual case where the JVM failed due to this bug?

I am not sure if we should have the assert. I think the assert will fail for cases like this:

```
./share/classfile/javaClasses.cpp:
log_error(class)("Mismatch JDK version for field: %s type: %s",
     name_symbol->as_C_string(), signature_symbol->as_C_string());
```

`as_C_string()` returns a string in UTF8 encoding. If the `name_symbol` contains non-ascii characters, the string may not be compatible with the JVM's current locale.

Klass::external_name() has a similar problem:

```
./share/classfile/verifier.cpp:
    log_info(verification)("Verifying class %s with old format",
                            klass->external_name());
```

Maybe it's a good idea to have the assert so we can fix these, but there are at least a few dozen cases in our code.

So I would suggest leaving the assert out of this PR, and perhaps do it in a separate RFE.

Reply:
I agree, lets keep this fix focused on one issue.

Reply:
Posix does potentially define a few other errors cases but the encoding error is what we ourselves document.

This is a good point @iklam. If  non-ascii class names could lead to problems then UL is far more fragile than I thought it to be! But I still think the assert is a good idea to detect this problem during development and testing as presumably during development and testing we don't want log messages to be silently dropped because of this encoding problem. That said we need to ensure we don't have existing tests that would cause this to trigger in the CI testing.

Reply:
Before, the VM would assumedly fail with a confusing native OOM. I cannot see how the subsequent malloc could have succeeded. So arguably, an assert is not worse, but clearer.

OTOH, this is "only" logging. We could just write something like "ENCODINGERROR" into the message buffer instead, to trip people off. Better than just to leave out the printed string.

Reply:
The other calls to `os::vsnprintf` does the same check and assert. I assume that we should have seen assertions failing by 
now, if the non-ASCII/non-compliant locale issue actually occurred.

@iklam, your link to StackOverflow links to this in-depth answer: https://groups.google.com/g/comp.std.c/c/lIvkxXr5_wE/m/aVuIpm52ToQJ (read James Kuyper's message). The TL;DR is that this is an issue exactly with `%ls` and `%lc` which are used for support with `wchar_t`.

Example of other site where `os::vsnprintf` is checked.

```
// logTagSet.cpp:130
  // Check that string fits in buffer; resize buffer if necessary
  int ret;
  if (prefix_len < vwrite_buffer_size) {
    ret = os::vsnprintf(buf + prefix_len, sizeof(buf) - prefix_len, fmt, args);
  } else {
    // Buffer too small. Just call printf to find out the length for realloc below.
    ret = os::vsnprintf(nullptr, 0, fmt, args);
  }

  assert(ret >= 0, "Log message buffer issue");
  if (ret < 0) {
    // Error, just log contents in buf.
    log(level, buf);
    log(level, "Log message buffer issue");
    va_end(saved_args);
    return;
  }
```

@tstuefe, I agree, and so does the example above. I will be adding a log message to this.

Reply:
I ran with `java -Xlog:all=debug --version`.  `vsnprintf()` was called 8495 times in `LogTagSet::vwrite()` and 746 times in `LogMessageBuffer::vwrite()`. So I agree with @jdksjolen that having an assert in the new code doesn't make the problem any worse than it's today.

We don't see the assert for a variety of reasons
- The VM currently doesn't use `%ls` and `%lc` in unified logging.
- Most people run Java under UTF8 locales.
- Most Java class/method/field names are ascii-only.

So we can probably leave things as is.